rayon-core/src/lib.rs

```diff
@@ -58,6 +59,243 @@ pub use future::RayonFuture;
+/// `set_num_threads()` method for details][snt]).
+///
+/// [snt]: struct.Configuration.html#method.set_num_threads
 pub fn current_num_threads() -> usize {
     ::registry::Registry::current_num_threads()
 }
+
+/// Contains the rayon thread pool configuration.
+pub struct Configuration {
+    /// The number of threads in the rayon thread pool.
+    /// If zero will use the RAYON_RS_NUM_CPUS environment variable.
+    /// If RAYON_RS_NUM_CPUS is invalid or zero will use the default.
+    num_threads: usize,
+
+    /// Custom closure, if any, to handle a panic that we cannot propagate
+    /// anywhere else.
+    panic_handler: Option<Box<PanicHandler>>,
+
+    /// Closure to compute the name of a thread.
+    get_thread_name: Option<Box<FnMut(usize) -> String>>,
+
+    /// The stack size for the created worker threads
+    stack_size: Option<usize>,
+
+    /// Closure invoked on worker thread start.
+    start_handler: Option<Box<StartHandler>>,
+
+    /// Closure invoked on worker thread exit.
+    exit_handler: Option<Box<ExitHandler>>,
+}
+
+/// The type for a panic handling closure. Note that this same closure
+/// may be invoked multiple times in parallel.
+type PanicHandler = Fn(Box<Any + Send>) + Send + Sync;
+
+/// The type for a closure that gets invoked when a thread starts. The
+/// closure is passed the index of the thread on which it is invoked.
+/// Note that this same closure may be invoked multiple times in parallel.
+type StartHandler = Fn(usize) + Send + Sync;
+
+/// The type for a closure that gets invoked when a thread exits. The
+/// closure is passed the index of the thread on which is is invoked.
+/// Note that this same closure may be invoked multiple times in parallel.
+type ExitHandler = Fn(usize) + Send + Sync;
+
+impl Configuration {
+    /// Creates and return a valid rayon thread pool configuration, but does not initialize it.
+    pub fn new() -> Configuration {
+        Configuration {
+            num_threads: 0,
+            get_thread_name: None,
+            panic_handler: None,
+            stack_size: None,
+            start_handler: None,
+            exit_handler: None,
+        }
+    }
+
+    /// Get the number of threads that will be used for the thread
+    /// pool. See `set_num_threads` for more information.
+    fn num_threads(&self) -> usize {
+        if self.num_threads > 0 {
+            self.num_threads
+        } else {
+            match env::var("RAYON_RS_NUM_CPUS").ok().and_then(|s| usize::from_str(&s).ok()) {
+                Some(x) if x > 0 => x,
+                _ => num_cpus::get(),
+            }
+        }
+    }
+
+    /// Get the thread name for the thread with the given index.
+    fn thread_name(&mut self, index: usize) -> Option<String> {
+        self.get_thread_name.as_mut().map(|c| c(index))
+    }
+
+    /// Set a closure which takes a thread index and returns
+    /// the thread's name.
+    pub fn set_thread_name<F>(mut self, closure: F) -> Self
```

Comment:
Now that the public methods are all setters, the `set_` prefix doesn't feel idiomatic.  Consider the standard library builders like [`Command`](https://doc.rust-lang.org/std/process/struct.Command.html) and [`thread::Builder`](https://doc.rust-lang.org/std/thread/struct.Builder.html) -- their methods are just named for the thing they're setting.

(But of course, we'd need to rename the private getters.)